src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionExtensions.cs

```diff
@@ -10,69 +10,79 @@ namespace Microsoft.AspNetCore.SignalR.Client
 {
     public static partial class HubConnectionExtensions
     {
-        private static void On(this HubConnection hubConnetion, string methodName, Type[] parameterTypes, Action<object[]> handler)
+        private static IDisposable On(this HubConnection hubConnetion, string methodName, Type[] parameterTypes, Action<object[]> handler)
         {
-            hubConnetion.On(methodName, parameterTypes, (parameters) =>
+            return hubConnetion.On(methodName, parameterTypes, (parameters, state) =>
             {
-                handler(parameters);
+                var currentHandler = (Action<object[]>)state;
+                currentHandler(parameters);
                 return Task.CompletedTask;
-            });
+            }, handler);
         }
 
-        public static void On(this HubConnection hubConnection, string methodName, Action handler)
+        public static IDisposable On(this HubConnection hubConnetion, string methodName, Type[] parameterTypes, Func<object[], Task> handler)
```

Comment:
Can you move this to the bottom of this file (i.e. as the last method)? This separates closely related extension methods.